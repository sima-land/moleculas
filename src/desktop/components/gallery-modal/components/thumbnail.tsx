import type { MouseEventHandler } from 'react';
import { MediaType } from '../types';
import { ImageOverlay } from './image-overlay';
import PlaySVG from '@sima-land/ui-quarks/icons/24x24/Filled/Play';
import classNames from 'classnames/bind';
import styles from './thumbnail.module.scss';
import AllRoundSVG from '@sima-land/ui-quarks/icons/40x40/Filled/Round360';

const cx = classNames.bind(styles);

export interface ThumbnailProps {
  type: MediaType;
  src?: string;
  alt?: string;
  checked?: boolean;
  onClick?: MouseEventHandler;
  className?: string;
}

/**
 * Превью медиа.
 * @param props Свойства.
 * @return Элемент.
 */
export const Thumbnail = ({ type, src, alt, checked, onClick, className }: ThumbnailProps) => (
  <button
    className={cx('root', type !== 'image' && 'iconic', { checked }, className)}
    onClick={onClick}
    data-testid='gallery-modal:thumbnail'
  >
    {type === '360' && <AllRoundSVG fill='currentColor' />}
    {type === 'video' && <PlaySVG width={32} height={32} fill='currentColor' />}
    {src && (
      <ImageOverlay className={cx('image-overlay')}>
        <img src={src} alt={alt} />
      </ImageOverlay>
    )}
  </button>
);
