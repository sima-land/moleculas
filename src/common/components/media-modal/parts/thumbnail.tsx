import { createContext, useContext, MouseEventHandler } from 'react';
import { ImageOverlay } from '../../../../desktop/components/gallery-modal/components/image-overlay';
import PlaySVG from '@sima-land/ui-quarks/icons/24x24/Filled/Play';
import AllRoundSVG from '@sima-land/ui-quarks/icons/40x40/Filled/Viewing360Perspective';
import { Price } from '@sima-land/ui-nucleons/price';
import { useImageStub } from '../../../hooks';
import classNames from 'classnames/bind';
import styles from './thumbnail.module.scss';
import { ImgStub } from '../../img-stub';

export type ThumbnailSize = 's' | 'l';

export interface ThumbnailProps {
  size?: ThumbnailSize;
  type: 'icon-360' | 'icon-video' | 'preview-image' | 'preview-video';
  src?: string;
  alt?: string;
  checked?: boolean;
  onClick?: MouseEventHandler;
  className?: string;
  loading?: boolean;
  title?: string;
  price?: number;
  currency?: string;
}

export const ThumbnailContext = createContext<Pick<ThumbnailProps, 'size'>>({});

const cx = classNames.bind(styles);

/**
 * Компонент превью медиа.
 * @param props Свойства.
 * @return Элемент.
 */
export function Thumbnail({
  size: sizeFromProps,
  type,
  src,
  alt,
  checked,
  onClick,
  className,
  loading,
  title,
  price,
  currency,
}: ThumbnailProps) {
  const { size: sizeFromContext } = useContext(ThumbnailContext);
  const { failed, handleError } = useImageStub(src);

  const size = sizeFromProps ?? sizeFromContext ?? 'l';

  const rootClassName = cx(
    'root',
    `size-${size}`,
    type?.startsWith('icon-') && 'iconic',
    checked && 'checked',
    loading && 'loading',
    className,
  );

  if (loading) {
    return (
      <button className={rootClassName} onClick={onClick} data-testid='thumbnail'>
        <span className={cx('content')} />
        {type === 'preview-video' && <span className={cx('info')} />}
      </button>
    );
  }

  return (
    <button className={rootClassName} onClick={onClick} data-testid='thumbnail'>
      <span className={cx('content')}>
        {type?.startsWith('preview-') && (
          <ImageOverlay className={cx('image')}>
            {failed && <ImgStub className={cx('stub')} />}
            {!failed && <img src={src} alt={alt} onError={handleError} />}
          </ImageOverlay>
        )}

        {!failed && type === 'preview-video' && <span className={cx('play')}></span>}

        {type === 'icon-360' && <AllRoundSVG fill='currentColor' />}

        {type === 'icon-video' && <PlaySVG width={32} height={32} fill='currentColor' />}
      </span>

      {(title || price) && (
        <span className={cx('info')}>
          {title && <span className={cx('title')}>{title}</span>}

          {price && <Price className={cx('price')} value={price} currencyGrapheme={currency} />}
        </span>
      )}
    </button>
  );
}
