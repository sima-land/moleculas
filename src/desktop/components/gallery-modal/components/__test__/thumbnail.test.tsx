import React from 'react';
import { mount } from 'enzyme';
import { Thumbnail } from '../thumbnail';
import { act } from 'react-dom/test-utils';

describe('Thumbnail', () => {
  it('should render 360 icon', () => {
    const spy = jest.fn();

    const wrapper = mount(<Thumbnail type='360' onClick={spy} />);

    expect(wrapper).toMatchSnapshot();

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find('button').simulate('click');
    });

    expect(spy).toBeCalledTimes(1);
  });

  it('should render video icon', () => {
    const spy = jest.fn();

    const wrapper = mount(<Thumbnail type='video' onClick={spy} />);

    expect(wrapper).toMatchSnapshot();

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find('button').simulate('click');
    });

    expect(spy).toBeCalledTimes(1);
  });

  it('should render image', () => {
    const spy = jest.fn();

    const wrapper = mount(<Thumbnail type='image' src='https://image.com/1' onClick={spy} />);

    expect(wrapper).toMatchSnapshot();

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find('button').simulate('click');
    });

    expect(spy).toBeCalledTimes(1);
  });

  it('should render 360 checked icon', () => {
    const spy = jest.fn();

    const wrapper = mount(<Thumbnail checked type='360' onClick={spy} />);

    expect(wrapper).toMatchSnapshot();
  });
});
