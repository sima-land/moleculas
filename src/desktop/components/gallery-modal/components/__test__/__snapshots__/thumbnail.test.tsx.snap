// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail should render 360 checked icon 1`] = `
<Thumbnail
  checked={true}
  onClick={[MockFunction]}
  type="360"
>
  <button
    className="root iconic checked"
    data-testid="gallery-modal:thumbnail"
    onClick={[MockFunction]}
  >
    <ForwardRef(Svg360)
      fill="currentColor"
    >
      <svg
        fill="currentColor"
        height={40}
        viewBox="0 0 40 40"
        width={40}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M30.63 17.878c.574-.839.861-2.13.861-3.876 0-1.686-.296-2.967-.889-3.842-.587-.875-1.458-1.312-2.61-1.312-1.186 0-2.063.419-2.633 1.257-.57.835-.854 2.133-.854 3.897 0 1.69.292 2.969.875 3.835.588.866 1.458 1.299 2.611 1.299 1.19 0 2.07-.42 2.639-1.258Zm-3.712-1.237c-.21-.534-.314-1.413-.314-2.639 0-1.226.107-2.106.32-2.639.215-.537.57-.806 1.067-.806.488 0 .841.268 1.06.806.223.534.335 1.413.335 2.639 0 1.212-.11 2.087-.328 2.625-.22.538-.575.807-1.067.807-.501 0-.859-.265-1.073-.793Zm-11.758-5.4c0 .624-.189 1.155-.567 1.592-.379.438-.91.738-1.593.902v.041c.807.1 1.417.347 1.832.739.415.387.622.911.622 1.572 0 .962-.349 1.711-1.046 2.249-.697.533-1.693.8-2.987.8-1.085 0-2.046-.18-2.885-.54v-1.798c.388.196.814.355 1.278.478.465.123.926.185 1.381.185.698 0 1.213-.119 1.545-.355.333-.238.5-.618.5-1.142 0-.47-.192-.802-.575-.998-.383-.196-.993-.294-1.832-.294h-.759v-1.62h.773c.774 0 1.34-.1 1.695-.301.36-.205.54-.554.54-1.046 0-.756-.474-1.135-1.422-1.135-.328 0-.663.055-1.005.164-.337.11-.713.299-1.128.568L8.55 9.846c.911-.657 1.998-.985 3.26-.985 1.035 0 1.85.21 2.448.63.601.419.902 1.002.902 1.75Zm1.333 3.513c0-1.978.417-3.45 1.251-4.416.839-.966 2.092-1.45 3.76-1.45.57 0 1.016.035 1.34.103v1.689a5.472 5.472 0 0 0-1.203-.137c-.725 0-1.317.11-1.778.328a2.21 2.21 0 0 0-1.025.97c-.228.43-.362 1.038-.403 1.826h.088c.452-.775 1.174-1.162 2.167-1.162.894 0 1.593.28 2.1.84.505.561.758 1.336.758 2.325 0 1.066-.3 1.912-.902 2.536-.602.62-1.436.93-2.502.93-.739 0-1.384-.171-1.935-.513-.547-.342-.97-.84-1.271-1.497-.297-.656-.445-1.447-.445-2.372Zm3.61 2.693c.45 0 .797-.15 1.039-.45.241-.306.362-.74.362-1.3 0-.487-.114-.87-.342-1.148-.223-.283-.563-.424-1.018-.424-.429 0-.796.139-1.101.417-.3.278-.451.602-.451.97 0 .543.141 1.001.424 1.375.287.373.649.56 1.087.56Zm14.923-3.141a1 1 0 1 0-1.052 1.701c1.155.714 1.902 1.45 2.362 2.2.456.74.664 1.543.664 2.45 0 1.704-1.251 3.48-3.813 4.979-2.526 1.479-6.123 2.552-10.27 2.9a1 1 0 0 0 .167 1.992c4.365-.366 8.268-1.5 11.114-3.166 2.81-1.645 4.802-3.95 4.802-6.706 0-1.236-.291-2.41-.96-3.498-.664-1.079-1.667-2.02-3.014-2.852Zm-29 2.044a1 1 0 1 0-1.052-1.702c-1.31.81-2.316 1.652-2.99 2.644C1.296 18.307 1 19.412 1 20.656c0 2.755 1.992 5.06 4.802 6.706 2.332 1.364 5.373 2.373 8.797 2.894l-1.746 1.48a1 1 0 1 0 1.294 1.526l3.5-2.967a1 1 0 0 0 .115-1.41l-3-3.533a1 1 0 0 0-1.524 1.294l1.343 1.583c-3.09-.503-5.765-1.42-7.768-2.593C4.25 24.136 3 22.36 3 20.656c0-.899.204-1.6.64-2.24.45-.666 1.195-1.33 2.386-2.066Z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(Svg360)>
  </button>
</Thumbnail>
`;

exports[`Thumbnail should render 360 icon 1`] = `
<Thumbnail
  onClick={[MockFunction]}
  type="360"
>
  <button
    className="root iconic"
    data-testid="gallery-modal:thumbnail"
    onClick={[MockFunction]}
  >
    <ForwardRef(Svg360)
      fill="currentColor"
    >
      <svg
        fill="currentColor"
        height={40}
        viewBox="0 0 40 40"
        width={40}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M30.63 17.878c.574-.839.861-2.13.861-3.876 0-1.686-.296-2.967-.889-3.842-.587-.875-1.458-1.312-2.61-1.312-1.186 0-2.063.419-2.633 1.257-.57.835-.854 2.133-.854 3.897 0 1.69.292 2.969.875 3.835.588.866 1.458 1.299 2.611 1.299 1.19 0 2.07-.42 2.639-1.258Zm-3.712-1.237c-.21-.534-.314-1.413-.314-2.639 0-1.226.107-2.106.32-2.639.215-.537.57-.806 1.067-.806.488 0 .841.268 1.06.806.223.534.335 1.413.335 2.639 0 1.212-.11 2.087-.328 2.625-.22.538-.575.807-1.067.807-.501 0-.859-.265-1.073-.793Zm-11.758-5.4c0 .624-.189 1.155-.567 1.592-.379.438-.91.738-1.593.902v.041c.807.1 1.417.347 1.832.739.415.387.622.911.622 1.572 0 .962-.349 1.711-1.046 2.249-.697.533-1.693.8-2.987.8-1.085 0-2.046-.18-2.885-.54v-1.798c.388.196.814.355 1.278.478.465.123.926.185 1.381.185.698 0 1.213-.119 1.545-.355.333-.238.5-.618.5-1.142 0-.47-.192-.802-.575-.998-.383-.196-.993-.294-1.832-.294h-.759v-1.62h.773c.774 0 1.34-.1 1.695-.301.36-.205.54-.554.54-1.046 0-.756-.474-1.135-1.422-1.135-.328 0-.663.055-1.005.164-.337.11-.713.299-1.128.568L8.55 9.846c.911-.657 1.998-.985 3.26-.985 1.035 0 1.85.21 2.448.63.601.419.902 1.002.902 1.75Zm1.333 3.513c0-1.978.417-3.45 1.251-4.416.839-.966 2.092-1.45 3.76-1.45.57 0 1.016.035 1.34.103v1.689a5.472 5.472 0 0 0-1.203-.137c-.725 0-1.317.11-1.778.328a2.21 2.21 0 0 0-1.025.97c-.228.43-.362 1.038-.403 1.826h.088c.452-.775 1.174-1.162 2.167-1.162.894 0 1.593.28 2.1.84.505.561.758 1.336.758 2.325 0 1.066-.3 1.912-.902 2.536-.602.62-1.436.93-2.502.93-.739 0-1.384-.171-1.935-.513-.547-.342-.97-.84-1.271-1.497-.297-.656-.445-1.447-.445-2.372Zm3.61 2.693c.45 0 .797-.15 1.039-.45.241-.306.362-.74.362-1.3 0-.487-.114-.87-.342-1.148-.223-.283-.563-.424-1.018-.424-.429 0-.796.139-1.101.417-.3.278-.451.602-.451.97 0 .543.141 1.001.424 1.375.287.373.649.56 1.087.56Zm14.923-3.141a1 1 0 1 0-1.052 1.701c1.155.714 1.902 1.45 2.362 2.2.456.74.664 1.543.664 2.45 0 1.704-1.251 3.48-3.813 4.979-2.526 1.479-6.123 2.552-10.27 2.9a1 1 0 0 0 .167 1.992c4.365-.366 8.268-1.5 11.114-3.166 2.81-1.645 4.802-3.95 4.802-6.706 0-1.236-.291-2.41-.96-3.498-.664-1.079-1.667-2.02-3.014-2.852Zm-29 2.044a1 1 0 1 0-1.052-1.702c-1.31.81-2.316 1.652-2.99 2.644C1.296 18.307 1 19.412 1 20.656c0 2.755 1.992 5.06 4.802 6.706 2.332 1.364 5.373 2.373 8.797 2.894l-1.746 1.48a1 1 0 1 0 1.294 1.526l3.5-2.967a1 1 0 0 0 .115-1.41l-3-3.533a1 1 0 0 0-1.524 1.294l1.343 1.583c-3.09-.503-5.765-1.42-7.768-2.593C4.25 24.136 3 22.36 3 20.656c0-.899.204-1.6.64-2.24.45-.666 1.195-1.33 2.386-2.066Z"
          fillRule="evenodd"
        />
      </svg>
    </ForwardRef(Svg360)>
  </button>
</Thumbnail>
`;

exports[`Thumbnail should render image 1`] = `
<Thumbnail
  onClick={[MockFunction]}
  src="https://image.com/1"
  type="image"
>
  <button
    className="root"
    data-testid="gallery-modal:thumbnail"
    onClick={[MockFunction]}
  >
    <ImageOverlay
      className="image-overlay"
    >
      <span
        className="root image-overlay"
      >
        <img
          src="https://image.com/1"
        />
      </span>
    </ImageOverlay>
  </button>
</Thumbnail>
`;

exports[`Thumbnail should render video icon 1`] = `
<Thumbnail
  onClick={[MockFunction]}
  type="video"
>
  <button
    className="root iconic"
    data-testid="gallery-modal:thumbnail"
    onClick={[MockFunction]}
  >
    <ForwardRef(PlaySVG)
      fill="currentColor"
      height={32}
      width={32}
    >
      <svg
        fill="currentColor"
        height={32}
        viewBox="0 0 24 24"
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 19.29V4.71a1 1 0 0 1 1.49-.872l12.96 7.29a1 1 0 0 1 0 1.744l-12.96 7.29A1 1 0 0 1 5 19.29Z"
        />
      </svg>
    </ForwardRef(PlaySVG)>
  </button>
</Thumbnail>
`;
